### PR TITLE
[MPS] Fix GPU hang in scatter_reduce/index_reduce with NaN

### PR DESCRIPTION
**Pull request by BenjaminDEMAILLE**
`atomic_compare_exchange_weak_explicit` on `::metal::atomic<float>` compares its
expected operand as a float, so once a NaN is stored the exchange never succeeds
and the CAS loop in `atomic_binary_op_helper`, which implements amax/amin/prod,
spins forever. `scatter_reduce_` and `index_reduce_` hang the GPU rather than
returning a wrong value or raising. Isolating the loop with
`torch.mps.compile_shader` and counting iterations, with a NaN already stored:

```
cas_float  iters= 100001 (guard tripped, never converges)
cas_bits   iters=      1
```

The fix does the exchange on the bit pattern through `atomic<uint>`, as the
uint-backed overload in the same header already does for half and bfloat. sum and
mean use atomic_fetch_add and were unaffected.

With the hang gone, `index_reduce_` still disagreed with CPU: `IndexReduceOp`
called unqualified min/max, which resolve to `metal::` and drop NaN, while the
equivalent functors in ScatterGather.metal qualify them as `::c10::metal::`. MPS
returned -inf, inf or 0.0 where CPU returns nan.

Test Plan:

```
python test/test_mps.py TestMPS.test_scatter_reduce_nan TestMPS.test_index_reduce_nan -v
python test/test_mps.py TestMPS TestGatherScatter TestAdvancedIndexing -q
```

The new tests cover amax/amin/prod/sum/mean, both values of include_self and
float32/float16/bfloat16. The suite is 1660 tests with one failure,
test_mps_allocator_module, which asserts nothing is allocated when it starts and
fails identically with these tests removed.

This PR was authored with the help of an AI assistant (Claude).

### CODE REVIEW

**Comment by BenjaminDEMAILLE** [comment]
Drafted with Claude Code, reviewed by me before posting (per `AI_POLICY.md`):

> @malfet the short version: `atomic_compare_exchange_weak_explicit` on `::metal::atomic<float>` compares its expected operand as a float, so once a NaN is stored the exchange never succeeds and the CAS loop behind `amax`/`amin`/`prod` spins forever. `scatter_reduce_` and `index_reduce_` hang the GPU rather than returning anything. Isolating the loop with `compile_shader` and counting iterations gives 100001 (guard tripped) against 1 for the bitwise variant. The fix is the same approach the uint-backed overload already uses for half and bfloat.

**Review by Isalia20 — COMMENTED** [review]
Doing this:
```
import torch

src_cpu = torch.tensor([float('nan'), 2.0, -3.0, float('nan')])
idx_cpu = torch.tensor([0, 1, 1, 2])
for reduce_str in ["amax", "amin", "prod", "mean"]:
   for include_self in [False, True]:
         out_cpu = torch.zeros(3)
         out_mps = torch.zeros(3, device='mps')
         out_cpu.index_reduce_(0, idx_cpu, src_cpu, reduce_str, include_self=include_self)
         out_mps.index_reduce_(0, idx_cpu.to('mps'), src_cpu.to('mps'), reduce_str,
                              include_self=include_self)
         torch.testing.assert_close(out_cpu, out_mps.cpu(), equal_nan=True)

```

Doesn't cause a hang, though values are incorrect. Can you give a reproducer where it hangs?

File: aten/src/ATen/native/mps/kernels/Indexing.metal [hunk 1]
```diff
@@ -287,14 +287,16 @@ struct IndexReduceOp {
...
-    return min(a, b);
+    return ::c10::metal::min(a, b);
   }
 
   template <typename T>
   static T amax(T a, T b) {
-    return max(a, b);
+    return ::c10::metal::max(a, b);
   }
 };
```
Comment (Isalia20) [inline aten/src/ATen/native/mps/kernels/Indexing.metal]: I think this should be a different PR, but I'm a bit confused. This:
```
import torch

a = torch.randn(4, 4)
a[0, 0] = float('nan')

print(torch.amin(a, 1))
```

returns 'nan' so wonder why this works if the comment says that min drop Nan

**Comment by BenjaminDEMAILLE** [comment]
I dug into this and I think it answers your question, though not in the direction you expected. Splitting the change doesn't work here.

Your example doesn't hang because no NaN ever gets stored. `IndexReduceOp` calls `metal::max`, which drops them, hence the `-inf` you're seeing instead of `nan`.

The hang is in `scatter_reduce_`. Its kernel uses `::c10::metal::max`, which does propagate NaN. Repro: `scatter_reduce_(0, idx, src, "amax", include_self=False)` with a NaN in `src`. Just ran it again and it hangs every time.

So the two fixes are coupled. If we only fix NaN propagation in `index_reduce_`, we introduce the hang there. Splitting them would land a patch that breaks things.

On the confusion: the comment is about `metal::min`/`metal::max`, the shader functions, not `torch.amin`. Happy to reword it if that's clearer.

File: aten/src/ATen/native/mps/kernels/Indexing.metal (hunk 1 above)
Comment (BenjaminDEMAILLE) [reply aten/src/ATen/native/mps/kernels/Indexing.metal]: Reworded — both namespaces are in scope here, so the unqualified names hit the Metal builtins.

**Comment by BenjaminDEMAILLE** [comment]
> Doing this:
> 
> ```
> import torch
> 
> src_cpu = torch.tensor([float('nan'), 2.0, -3.0, float('nan')])
> idx_cpu = torch.tensor([0, 1, 1, 2])
> for reduce_str in ["amax", "amin", "prod", "mean"]:
>    for include_self in [False, True]:
>          out_cpu = torch.zeros(3)
>          out_mps = torch.zeros(3, device='mps')
>          out_cpu.index_reduce_(0, idx_cpu, src_cpu, reduce_str, include_self=include_self)
>          out_mps.index_reduce_(0, idx_cpu.to('mps'), src_cpu.to('mps'), reduce_str,
>                               include_self=include_self)
>          torch.testing.assert_close(out_cpu, out_mps.cpu(), equal_nan=True)
> ```
> 
> Doesn't cause a hang, though values are incorrect. Can you give a reproducer where it hangs?

The example can't hang: idx=[0,1,1,2] never writes twice to a slot holding a NaN.
Slots 0 and 2 get one write each, slot 1 gets two but no NaN. A single CAS
succeeds on the first try even when it stores a NaN.

Two writes are needed, the first being the NaN. And index_reduce_ can't produce
that today for a second reason: IndexReduceOp::amax calls metal::max, which
drops the NaN, so no NaN is ever stored. That's the -inf you see instead of nan.
The two bugs mask each other.

scatter_reduce_ has no such mask, ScatterGather.metal already uses
::c10::metal::max, so it hangs today on main:

    vals = torch.tensor([float("nan"), -1.39, 2.37, 4.0, -8.15, 5.01, 1.36], device="mps")
    idx = torch.tensor([0, 2, 4, 1, 4, 0, 3], device="mps")
    out = torch.zeros(5, device="mps")
    out.scatter_reduce_(0, idx, vals, "amax", include_self=False)
    torch.mps.synchronize()

Slot 0 gets writes from positions 0 and 5, and position 0 is the NaN.

This is why the two changes ship together. Fixing the propagation alone turns
index_reduce_ into the same shape as scatter_reduce_, so idx=[0,0] with
src=[nan, 5.0] would then hang. I measured the CAS loop in isolation with
compile_shader, same loop, only the max differing: metal::max converges in 3
iterations, c10::metal::max never converges (stopped at 100k).

amin and prod behave identically. sum and mean are unaffected, they use
atomic_fetch_add.
